File: JavaSimpleBackend/src/main/java/com/atesti/entity/RadniNalog.java
package com.atesti.entity;

import com.fasterxml.jackson.annotation.JsonBackReference;
import com.fasterxml.jackson.annotation.JsonManagedReference;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import jakarta.persistence.*;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "radni_nalozi")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class RadniNalog {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false, unique = true)
    private String brojNaloga;

    @Column(nullable = false)
    private LocalDateTime datum;

    @Column(nullable = false)
    private String objekt;

    @Builder.Default
    private Boolean fakturirano = false;

    @Builder.Default
    private Boolean zavrseno = false;

    @Column(columnDefinition = "TEXT")
    private String opis;

    private String brojPonude;
    private String brojRacuna;
    private String narudzbenica;
    private String ugovor;

    @Column(columnDefinition = "TEXT")
    private String aktivnosti;

    private String pdfUrl;

    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "narucitelj_id")
    @JsonBackReference
    private Narucitelj narucitelj;

    @Column(name = "narucitelj_id", insertable = false, updatable = false)
    private Long naruciteljId;

    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "assigned_user_id")
    private User assignedUser;

    @Column(name = "assigned_user_id", insertable = false, updatable = false)
    private Long assignedUserId;

    @OneToMany(mappedBy = "radniNalog", fetch = FetchType.LAZY, cascade = CascadeType.ALL)
    @JsonManagedReference("radniNalog-documents")
    @Builder.Default
    private List<Document> documents = new ArrayList<>();

    @OneToMany(mappedBy = "radniNalog", fetch = FetchType.LAZY, cascade = CascadeType.ALL)
    @JsonManagedReference("radniNalog-notes")
    @Builder.Default
    private List<Note> notes = new ArrayList<>();

    @CreationTimestamp
    private LocalDateTime createdAt;

    @UpdateTimestamp
    private LocalDateTime updatedAt;
}

File: backend/models/radni_nalog.js
module.exports = (sequelize, DataTypes) => {
  const RadniNalog = sequelize.define('radni_nalog', {
    id: {
      type: DataTypes.INTEGER,
      primaryKey: true,
      autoIncrement: true
    },
    brojNaloga: {
      type: DataTypes.STRING,
      allowNull: false,
      unique: true
    },
    datum: {
      type: DataTypes.DATE,
      allowNull: false
    },
    objekt: {
      type: DataTypes.STRING,
      allowNull: false
    },
    fakturirano: {
      type: DataTypes.BOOLEAN,
      defaultValue: false
    },
    zavrseno: {
      type: DataTypes.BOOLEAN,
      defaultValue: false
    },
    opis: {
      type: DataTypes.TEXT,
      allowNull: true
    },
    brojPonude: {
      type: DataTypes.STRING,
      allowNull: true
    },
    brojRacuna: {
      type: DataTypes.STRING,
      allowNull: true
    },
    narudzbenica: {
      type: DataTypes.STRING,
      allowNull: true
    },
    ugovor: {
      type: DataTypes.STRING,
      allowNull: true
    },
    aktivnosti: {
      type: DataTypes.TEXT,
      allowNull: true,
      get() {
        const rawValue = this.getDataValue('aktivnosti');
        if (!rawValue) {
          return [];
        }

        if (Array.isArray(rawValue)) {
          return rawValue;
        }

        if (typeof rawValue === 'string') {
          try {
            const parsed = JSON.parse(rawValue);
            return Array.isArray(parsed) ? parsed : [rawValue];
          } catch (error) {
            return [rawValue];
          }
        }

        return [];
      },
      set(value) {
        if (value === null || value === undefined) {
          this.setDataValue('aktivnosti', null);
          return;
        }

        if (Array.isArray(value)) {
          this.setDataValue('aktivnosti', JSON.stringify(value));
          return;
        }

        if (typeof value === 'string') {
          try {
            const parsed = JSON.parse(value);
            if (Array.isArray(parsed)) {
              this.setDataValue('aktivnosti', JSON.stringify(parsed));
              return;
            }
          } catch (error) {
          }
          this.setDataValue('aktivnosti', JSON.stringify([value]));
          return;
        }

        this.setDataValue('aktivnosti', JSON.stringify([]));
      }
    },
    pdfUrl: {
      type: DataTypes.STRING,
      allowNull: true
    },
    assignedUserId: {
      type: DataTypes.INTEGER,
      allowNull: true
    }
  }, {
    timestamps: true,
    tableName: 'radni_nalozi'
  });

  RadniNalog.associate = (models) => {
    RadniNalog.belongsTo(models.narucitelj, {
      foreignKey: 'narucitelj_id',
      as: 'narucitelj'
    });
    RadniNalog.hasMany(models.document, {
      foreignKey: 'radni_nalog_id',
      as: 'documents'
    });
    RadniNalog.hasMany(models.note, {
      foreignKey: 'radni_nalog_id',
      as: 'notes'
    });
    RadniNalog.belongsTo(models.user, {
      foreignKey: 'assignedUserId',
      as: 'assignedUser'
    });
  };

  return RadniNalog;
};
